tests/test_cmdline_date.py

```diff
@@ -0,0 +1,105 @@
+# =================================================================================================
+# Copyright (C) 2018-2019 University of Glasgow
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice,
+#    this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+# SPDX-License-Identifier: BSD-2-Clause
+# =================================================================================================
+import npt.util
+import unittest as ut
+import tempfile, shutil, pathlib
+from datetime import datetime, timedelta
+import time
+import sys 
+
+def program_name_idx():
+    for _idx, arg in enumerate(sys.argv):
+        if arg.find(__name__) >= 0 :
+            return _idx 
+    return 0
+
+class Test_Cmdline_Date(ut.TestCase):
+    def setUp(self):
+        # generate a temporary directory name
+        # remove actual directory - tool should a
+        self.rootdir = pathlib.Path(tempfile.mkdtemp())
+        self.argv = sys.argv[program_name_idx():]
+
```

Comment:
I don't understand why these tests need to mess with `sys.argv`. Why not pass a sample command line to the test function?

Reply:
This line is only picking up the name of the calling program. The test code is directly passing in arguments as parameters 
into command line parsing.  e.g. line 52,  65 etc

Reply:
Sure - but why does it need to store this between functions at all? You don't need an `argv` instance variable. It can be local to each function. As a general principle, minimise shared state.

Reply:
Okay will rewrite these for each test within each method. 
Since the setUp() / tearDown() functions are called for each new test_method() within a unitTest class, it would have reduced boilerplate -- which is why it was written like this  :)

Reply:
Sure - I tend to prefer boilerplate to implicit state, but maybe that's me.

Thanks for indulging my comments

Reply:
Not at all. I  did not think of it as implicit state  -- just as a means to reduce boilerplate.  I do agree, that minimising global state is good practice.  Done